HACKING.md

```diff
@@ -0,0 +1,39 @@
+`GSSAPI`, `KERBEROS_V4`, `DIGEST-MD5` | The `GSSAPI`, `KERBEROS_V4`, `DIGEST-MD5` mechanisms all require the set-up of an external authentication mechanism, either Kerberos or an authentication identity mapping. They furthermore require `root`-level access to `olcDatabase{0}config` to initiate the setup. 
+
+As is clear, in order to use SASL, `root`-level access to the `olcDatabase{0}config` database is required. If we are trying to acces this database in order to commit `root`-level credentials or information like schemas or the `top` object, this is a self-referential problem.
+
+In order to use TLS as the underlying mechanism for SASL `EXTERNAL`, one can only vaguely follow the [documentation](http://www.openldap.org/faq/data/cache/185.html) as it continues to refer to the `slapd.conf` schema for configuration, which is deprecated in the current OpenLDAP release. Furthermore, pointing the OpenLDAP daemon to certs requires `root`-level access to `olcDatabase{0}config`, which, again, is an endless problem if you are attempting to create an authentication scheme so you can access that database.
+
+The default out-of-the-box installation of OpenLDAP contains a `olcRootDN` manager for the `cn=config` database but does not list a password, which is problematic as it does not allow you to use the full authentication (`-x -D <managerCN> -w <password>`) and therefore does not allow for the user to use this set-up to add or remove records unless they are `UID 0`.
+
+### OpenLDAP Configuration Dump
+OpenLDAP exposes administrative tools like [`slapadd`](http://www.openldap.org/software/man.cgi?query=slapadd&apropos=0&sektion=0&manpath=OpenLDAP+2.4-Release&format=html) and [`slpacat`](http://www.openldap.org/software/man.cgi?query=slapcat&apropos=0&sektion=0&manpath=OpenLDAP+2.4-Release&format=html) which operate on OpenLDAP installations with the daemon not running. These tools are a promising work-around to the authentication nightmare above. The gist of this approach is as such: first, install OpenLDAP on a machine or container where you have `root` privileges (`UID 0` and `GID 0` are required due to how OpenLDAP generates authentication identities), create an OpenLDAP installation to your liking, stop the daemon (`slapd`), dump all of the OpenLDAP installation to `ldif` using `slapcat -n0`, then import all of the data using `slapadd -n0` on another installation. This approach does not seem to work as if there are no vestiges of the new `cn=config` configuration schema in the `etc/openldap` directory, the `slapadd` command looks for the old `slapd.conf` configuration schema and fails because it is deprecated. `slapadd` also cannot be used while the daemon is off to insert the records we want to `olcDatabase{0}config` as it is only capable of adding object, not editing or amending current objects. Editing the files by hand will trigger checksum errors and cause a corrupt installation. Therefore, this option was abandoned.
```

Comment:
work on the tone a little...  "nightmare" is probably not how we want to describe this.